apps/prairielearn/src/middlewares/AuthzAccessMismatch.tsx

```diff
@@ -112,42 +130,44 @@ const PermissionsTable = ({ permissions }: { permissions: PermissionData[] }) =>
       </tbody>
     </table>
   );
-};
+}
+function clearEffectiveUserCookies() {
+  removeCookieClient(['pl_requested_uid', 'pl2_requested_uid']);
+  removeCookieClient(['pl_requested_course_role', 'pl2_requested_course_role']);
+  removeCookieClient(['pl_requested_course_instance_role', 'pl2_requested_course_instance_role']);
+  removeCookieClient(['pl_requested_date', 'pl2_requested_date']);
+  setCookieClient(['pl_requested_data_changed', 'pl2_requested_data_changed'], 'true');
+  window.location.reload();
+}
 
 export function AuthzAccessMismatch({
-  errorMessage,
+  errorExplanation,
   oneOfPermissionKeys,
   authzData,
   authnUser,
   authzUser,
 }: {
-  errorMessage: string;
-  oneOfPermissionKeys: (keyof PageContext['authz_data'])[];
+  /**
+   * A sentence-like description of why the user can't access the page. If not provided,
+   * one will be generated from `oneOfPermissionKeys`.
+   */
+  errorExplanation?: string;
+  oneOfPermissionKeys: CheckablePermissionKeys[];
   authzData: PageContext['authz_data'];
   authnUser: StaffUser;
   authzUser: StaffUser | null;
 }) {
-  const clearEffectiveUserCookies = () => {
-    removeCookieClient(['pl_requested_uid', 'pl2_requested_uid']);
-    removeCookieClient(['pl_requested_course_role', 'pl2_requested_course_role']);
-    removeCookieClient(['pl_requested_course_instance_role', 'pl2_requested_course_instance_role']);
-    removeCookieClient(['pl_requested_date', 'pl2_requested_date']);
-    setCookieClient(['pl_requested_data_changed', 'pl2_requested_data_changed'], 'true');
-    window.location.reload();
-  };
-
-  const permissions: PermissionData[] = PermissionsMeta.map((permission) => {
+  const permissions: PermissionData[] = PERMISSIONS_META.map((permission) => {
     return {
       authnValue:
-        (authzData as any)['authn_' + permission.key] ??
-        (permission.type === 'string' ? '' : false),
+        authzData[`authn_${permission.key}`] ?? (permission.type === 'string' ? '' : false),
```

Comment:
We can use [template literal types](https://www.typescriptlang.org/docs/handbook/2/template-literal-types.html) to avoid the cast and give ourselves extra type safety!